Add new Meteoclimatic integration

adrianmo opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

New integration to support meteorological data from the [Meteoclimatic](https://www.meteoclimatic.net/) service. Meteoclimatic is a network of non-professional automatic real-time weather stations. The geographical scope of Meteoclimatic is mostly Spain, Portugal, and Andorra.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13786
- Link to Brands pull request: https://github.com/home-assistant/brands/pull/1711

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

adrianmo committed:
Add Meteoclimatic integration

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

adrianmo committed:
Update .coveragerg

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

homeassistant commented:
Hi @adrianmo, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#36906).

Once you do that we will be able to review and accept this pull request.

Thanks!

adrianmo committed:
Merge branch 'dev' into meteoclimatic

adrianmo committed:
Fix isort offenses

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

adrianmo committed:
Fix isort offenses in test files

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

MatthewFlamm commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,103 @@
+    return unload_ok
+
+
+class MeteoclimaticUpdater:
```
Since you have a single API call common to all entities, you could consider using DataUpdateCoordinator as in https://developers.home-assistant.io/docs/integration_fetching_data#coordinated-single-api-poll-for-data-for-all-entities

adrianmo replied:
Thanks for suggestion. I've refactored to use DataUpdateCoordinator.

MatthewFlamm commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,103 @@
+    station_code = entry.data[CONF_STATION_CODE]
+    client = MeteoclimaticClient()
+    hass.data[DOMAIN][station_code] = MeteoclimaticUpdater(client, station_code)
+    await hass.async_add_executor_job(hass.data[DOMAIN][station_code].update)
```
What happens if this fails?

adrianmo replied:
I'm now capturing the `MeteoclimaticError` in the `MeteoclimaticUpdater._update()` method.

adrianmo committed:
Refactor and switch to DataUpdateCoordinator

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

MatthewFlamm commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,136 @@
+
+    meteoclimatic_updater = MeteoclimaticUpdater(hass, station_code)
+    await meteoclimatic_updater.async_update()
+    meteoclimatic_data = meteoclimatic_updater.get_data()
```
It looks like this update and data check, and the coordinator refresh below are redundant.

adrianmo replied:
You are right. I just removed the coordinator refresh.

MatthewFlamm commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,136 @@
+        """Obtain the latest data from Meteoclimatic."""
+        try:
+            self._data = self._client.weather_at_station(self._station_code)
+        except MeteoclimaticError as exp:
```
You could instead reraise UpdateFailed instead of logging here, so that it takes advantage of limited logging in DataUpdateCoordinator

See https://github.com/home-assistant/core/blob/4aedafc73af5f7679b0544d77726be8192c5462a/homeassistant/helpers/update_coordinator.py#L148

adrianmo replied:
Done!

MatthewFlamm commented on homeassistant/components/meteoclimatic/sensor.py:
```diff
@@ -0,0 +1,127 @@
+    @property
+    def available(self):
+        """Return if state is available."""
+        return self._data is not None
```
You can use the `last_update_success` attribute of the coordinator, otherwise it looks like this evaluates as true if _any_ previous update succeeded.

adrianmo replied:
Done. Thanks for the suggestion!

MatthewFlamm commented on homeassistant/components/meteoclimatic/weather.py:
```diff
@@ -0,0 +1,151 @@
+    @property
+    def available(self):
+        """Return if state is available."""
+        return self._data is not None
```
See comment for sensor

adrianmo replied:
Done.

adrianmo committed:
Address code review comments

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

adrianmo commented:
@MatthewFlamm could you please have another look at the PR? Thank you!

MatthewFlamm commented on homeassistant/components/meteoclimatic/sensor.py:
```diff
@@ -0,0 +1,127 @@
+
+    async_add_entities(
+        [MeteoclimaticSensor(sensor_type, hass_data) for sensor_type in SENSOR_TYPES],
+        True,
```
I think you can set this second parameter to False since you have already requested the data in `async_setup_entry` in `__init.py__`.

adrianmo replied:
Updated!

MatthewFlamm commented on homeassistant/components/meteoclimatic/weather.py:
```diff
@@ -0,0 +1,151 @@
+    """Set up the Meteoclimatic weather platform."""
+    hass_data = hass.data[DOMAIN][entry.entry_id]
+
+    async_add_entities([MeteoclimaticWeather(hass_data)], True)
```
Same comment about second parameter.

adrianmo replied:
Updated!

MatthewFlamm reviewed (commented):
It is looking good to my eyes, but you will have to wait until someone more knowledgeable (and has write permissions) has time to review.

adrianmo committed:
set update_before_add arg to False

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

adrianmo commented:
Thanks @MatthewFlamm - I've made the changes suggested and updated the branch. Happy to make further changes until it's ready to be merged.

adrianmo commented:
Is there any other change request?

adrianmo commented:
This PR is still waiting for the follow-up review.

epenet reviewed (changes requested):
May I suggest that you add some extra common keys to the config_flow strings, as discussed on #40578?
At least for already_configured and unknown.

MatthewFlamm commented on homeassistant/components/meteoclimatic/const.py:
```diff
@@ -0,0 +1,134 @@
+}
+
+CONDITION_CLASSES = {
+    "clear-night": [Condition.moon, Condition.hazemoon],
```
A lot of these constants (all?) inside this dict are now defined in `homeassistant.components.weather`

MatthewFlamm replied:
Now that I think about this more, if you also use this inside the sensor platform, it may not optimal to use those constants for the sensor platform.

adrianmo replied:
@MatthewFlamm as suggested, I've updated the condition class keys to use the constants in `homeassistant.components.weather`. However, I'm not sure what you mean with your second comment. Can you please clarify that for me? Thanks.

MatthewFlamm replied:
Please disregard my second comment, it does not look like you use this constant in the sensor module.

epenet commented on homeassistant/components/meteoclimatic/const.py:
```diff
@@ -0,0 +1,134 @@
+    },
+    "rain": {
+        SENSOR_TYPE_NAME: "Rain",
+        SENSOR_TYPE_UNIT: "mm",
```
LENGTH_MILLIMETERS is available in homeassistant.const

adrianmo replied:
Done. Thank you.

adrianmo committed:
Use new const LENGTH_MILLIMETERS and rename PERCENTAGE const

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

adrianmo committed:
Use weather attr constants instead of string literals

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

adrianmo committed:
Use string references in config flow

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

adrianmo committed:
Merge branch 'dev' into meteoclimatic

adrianmo commented:
@epenet I have updated the `strings.json` file to use references to common keys. Thank you.

adrianmo committed:
Fix pylint W0707 offense

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

frenck commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,128 @@
+
+STATION_CODE_SCHEMA = vol.Schema({vol.Required(CONF_STATION_CODE): cv.string})
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema(vol.All(cv.ensure_list, [STATION_CODE_SCHEMA]))},
+    extra=vol.ALLOW_EXTRA,
+)
```
This is a new integration, it should not have a configuration schema.

frenck commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,128 @@
+    if conf is None:
+        return True
+
+    for station_code_conf in conf:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={"source": SOURCE_IMPORT}, data=station_code_conf.copy()
+            )
+        )
```
Import is used for migrating integrations from YAML -> configuration entries. Since this is a new integration, this can be removed.

frenck commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,128 @@
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+
+    _LOGGER.debug("meteoclimatic sensor platform loaded for %s", station_code)
```
This is already logged by the core.

frenck commented on homeassistant/components/meteoclimatic/const.py:
```diff
@@ -0,0 +1,152 @@
+    "temp_current": {
+        SENSOR_TYPE_NAME: "Temperature",
+        SENSOR_TYPE_UNIT: TEMP_CELSIUS,
+        SENSOR_TYPE_ICON: "mdi:thermometer",
+        SENSOR_TYPE_CLASS: DEVICE_CLASS_TEMPERATURE,
```
Either set a device class or icon, not both.

frenck commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,128 @@
+    return unload_ok
+
+
+class MeteoclimaticUpdater:
```
You actually don't need this class.

If you take `_update`, make data method return the actual data, and use that as the update method for the DataUpdateCoordinator, the coordinator would have/hold the data and pass it along.

You do not need a class to hold that, that is what the coordinator is for.

frenck commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,128 @@
+        hass,
+        _LOGGER,
+        name=f"Meteoclimatic Coordinator for {station_code}",
+        update_method=meteoclimatic_updater.async_update,
```
There should be a method passed in, that actually returns the data.

frenck commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,128 @@
+    meteoclimatic_hass_data = hass.data.setdefault(DOMAIN, {})
+    meteoclimatic_hass_data[entry.entry_id] = {
+        METEOCLIMATIC_COORDINATOR: coordinator,
+        METEOCLIMATIC_STATION_CODE: meteoclimatic_data.station.code,
+        METEOCLIMATIC_STATION_NAME: meteoclimatic_data.station.name,
```
This is part of the data set, there is no need to store these separately?

frenck commented on homeassistant/components/meteoclimatic/sensor.py:
```diff
@@ -0,0 +1,127 @@
+        self._type = sensor_type
+        self._unique_id = f"{hass_data[METEOCLIMATIC_STATION_CODE]}_{self._type}"
+        self._name = f"{hass_data[METEOCLIMATIC_STATION_NAME]} {SENSOR_TYPES[self._type][SENSOR_TYPE_NAME]}"
+        self._updater = hass_data[METEOCLIMATIC_UPDATER]
```
The coordinator can handle the updates and the data.

frenck commented on homeassistant/components/meteoclimatic/sensor.py:
```diff
@@ -0,0 +1,127 @@
+        self._updater = hass_data[METEOCLIMATIC_UPDATER]
+        self._coordinator = hass_data[METEOCLIMATIC_COORDINATOR]
+
+        self._data = None
```
The coordinator holds the data, this is not needed.

frenck commented on homeassistant/components/meteoclimatic/sensor.py:
```diff
@@ -0,0 +1,127 @@
+    )
+
+
+class MeteoclimaticSensor(Entity):
```
Extend from CoordinatorEntity, as it implements some basics for the data update coordinator.

frenck commented on homeassistant/components/meteoclimatic/sensor.py:
```diff
@@ -0,0 +1,127 @@
+        self._data = self._updater.get_data()
+        self.async_write_ha_state()
+
+    @property
+    def should_poll(self) -> bool:
+        """Entities do not individually poll."""
+        return False
```
When using the CoordinatorEntity, this can be removed.

frenck commented on homeassistant/components/meteoclimatic/sensor.py:
```diff
@@ -0,0 +1,127 @@
+        )
+        self._update_callback()
+
+    async def async_update(self):
+        """Schedule a custom update via the common entity update service."""
+        await self._coordinator.async_request_refresh()
```
When using the CoordinatorEntity, this can be removed.

frenck commented on homeassistant/components/meteoclimatic/sensor.py:
```diff
@@ -0,0 +1,127 @@
+        self.async_on_remove(
+            self._coordinator.async_add_listener(self._update_callback)
+        )
+        self._update_callback()
```
This should not be needed, as the coordinator can hold all the data already.

frenck commented on homeassistant/components/meteoclimatic/sensor.py:
```diff
@@ -0,0 +1,127 @@
+            ATTR_STATION_NAME: self._data.station.name if self._data else None,
+        }
+
+    async def async_added_to_hass(self) -> None:
+        """Set up a listener and load data."""
+        self.async_on_remove(
+            self._coordinator.async_add_listener(self._update_callback)
+        )
```
When using the CoordinatorEntity, this can be removed.

frenck commented on homeassistant/components/meteoclimatic/sensor.py:
```diff
@@ -0,0 +1,127 @@
+        self._unique_id = f"{hass_data[METEOCLIMATIC_STATION_CODE]}_{self._type}"
+        self._name = f"{hass_data[METEOCLIMATIC_STATION_NAME]} {SENSOR_TYPES[self._type][SENSOR_TYPE_NAME]}"
+        self._updater = hass_data[METEOCLIMATIC_UPDATER]
+        self._coordinator = hass_data[METEOCLIMATIC_COORDINATOR]
```
The coordinator is already in CoordinatorEntity (when used)

frenck commented on homeassistant/components/meteoclimatic/translations/ca.json:
```diff
@@ -0,0 +1,20 @@
+{
```
New languages should not be added here. Translations can be managed via Lokalised after this PR is merged.

frenck commented on homeassistant/components/meteoclimatic/strings.json:
```diff
@@ -0,0 +1,20 @@
+    "step": {
+      "user": {
+        "title": "Meteoclimatic",
+        "description": "Enter the Meteoclimatic station code (e.g. ESCAT4300000043206B)",
```
```suggestion
        "description": "Enter the Meteoclimatic station code (e.g., ESCAT4300000043206B)",
```

frenck commented on homeassistant/components/meteoclimatic/translations/es.json:
```diff
@@ -0,0 +1,20 @@
+{
+    "config": {
+        "abort": {
```
Please remove this additional language.

frenck commented on homeassistant/components/meteoclimatic/weather.py:
```diff
@@ -0,0 +1,151 @@
+"""Support for Meteoclimatic weather service."""
```
There is a sensor and weather platform in this PR. Please limit PRs to adding a single platform.

So either limit it to the sensor or to weather for this PR. Additional platforms can be added in a follow-up PR.

frenck reviewed (commented):
Some initial comments to get the review process going 👍

adrianmo commented:
Thanks for the review @frenck - I'll go through the comments and fix the issues shortly 👍

adrianmo committed:
Merge branch 'dev' into meteoclimatic

adrianmo committed:
Fix frenck comments

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

adrianmo committed:
Enable config from file

Signed-off-by: Adrián Moreno <adrian@morenomartinez.com>

adrianmo commented:
Hi @frenck, I've gone through all your comments and I've addressed them one by one. Major updates are:

- Refactored the data update logic and Weather entity to use the Coordinator framework
- Removed the Sensor platform (will be added in a separate PR)
- Removed the translations (will be added in a separate PR)

Even though this is a new integration, I wanted to let users configure it programmatically via config file as well. But if that's not acceptable, I can also get rid of it.

Thanks for your time.

adrianmo commented:
Hi @frenck, I've gone through all your comments and I've addressed them one by one. Major updates are:

- Refactored the data update logic and Weather entity to use the Coordinator framework
- Removed the Sensor platform (will be added in a separate PR)
- Removed the translations (will be added in a separate PR)

Even though this is a new integration, I wanted to let users configure it programmatically via config file as well. But if that's not acceptable, I can also get rid of it.

Thanks for your time.

frenck commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,87 @@
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in PLATFORMS
+            ]
+        )
+    )
```
```suggestion
    """Unload a config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```

frenck commented on homeassistant/components/meteoclimatic/config_flow.py:
```diff
@@ -0,0 +1,71 @@
+from homeassistant import config_entries
+
+from .const import CONF_STATION_CODE
+from .const import DOMAIN  # pylint: disable=unused-import
```
```suggestion
from .const import DOMAIN
```

frenck commented on homeassistant/components/meteoclimatic/config_flow.py:
```diff
@@ -0,0 +1,71 @@
+    """Handle a Meteoclimatic config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
```
This is no longer used. 
```suggestion
```

frenck commented on homeassistant/components/meteoclimatic/manifest.json:
```diff
@@ -0,0 +1,12 @@
+  ],
+  "codeowners": [
+    "@adrianmo"
+  ]
```
```suggestion
  ],
  "iot_class": "cloud_polling"
```

frenck commented on homeassistant/components/meteoclimatic/weather.py:
```diff
@@ -0,0 +1,92 @@
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
```
```suggestion
    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
```

frenck commented on tests/components/meteoclimatic/test_config_flow.py:
```diff
@@ -0,0 +1,134 @@
+def mock_setup():
+    """Prevent setup."""
+    with patch(
+        "homeassistant.components.meteoclimatic.async_setup",
+        return_value=True,
+    ), patch(
```
```suggestion
```

frenck reviewed (commented):
I think this is looking really good already!

If it is rebased onto the latest dev, the suggestions above should be able to apply (which cleans things up a little). If that is done I'll do a final pass and test.

All in all, good job! 👍

adrianmo committed:
Merge branch 'dev' into meteoclimatic

adrianmo committed:
Apply frenck's suggestions

Signed-off-by: Adrian Moreno <adrian@morenomartinez.com>

adrianmo commented:
@frenck Thanks for the review! I've made the suggested changes and brought the branch up to date with `dev`. Happy to make as many changes as needed to get this shipped 🚀

frenck commented on homeassistant/components/meteoclimatic/config_flow.py:
```diff
@@ -0,0 +1,69 @@
+            title=weather.station.name, data={CONF_STATION_CODE: station_code}
+        )
+
+    async def async_step_import(self, user_input):
+        """Import a config entry."""
+        return await self.async_step_user(user_input)
```
Nothing is using this step at this point :)
```suggestion
```

frenck commented on homeassistant/components/meteoclimatic/__init__.py:
```diff
@@ -0,0 +1,52 @@
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry):
```
```suggestion
async def async_unload_entry(hass: HomeAssistantType, entry: ConfigEntry) -> None:
```

adrianmo replied:
Is the function returning `None`? I've added `-> bool` instead, but let me know if I'm missing something.

frenck reviewed (changes requested):
2 small comments/suggestions left. Otherwise, it looks good to me. Tested it locally as well 👍 

Left some comments on the documentation PR to catch those up with this PR.

adrianmo committed:
Remove unused function and add return type hint

Signed-off-by: Adrian Moreno <adrian@morenomartinez.com>

adrianmo commented:
@frenck Thanks again for the thorough review! I've applied your suggestions and also updated the documentation PR  :)

adrianmo committed:
Remove import tests

Signed-off-by: Adrian Moreno <adrian@morenomartinez.com>

frenck reviewed (approved):
✅ This looks good to me. Tested locally and works.

I'm asking for a second opinion from another reviewer though 😄

janiversen commented on .coveragerc:
```diff
@@ -600,6 +600,9 @@ omit =
     homeassistant/components/meteo_france/weather.py
     homeassistant/components/meteoalarm/*
+    homeassistant/components/meteoclimatic/__init__.py
+    homeassistant/components/meteoclimatic/const.py
```
We should avoid adding files to .coveragerc, since it means they are not tested. Are you sure const.py and __init__.py are not tested?
you can run coverage locally with:
pytest --cov-report html --cov=../../../homeassistant/components/meteoclimatic .
from your testdirectiry (hence the .)
report is in directory htmlcov

balloob replied:
That's actually fine. We don't require these to be tested. It's opt-in.

janiversen reviewed (dismissed):
Looks good to me, please have a look at coverage for your next PR (see my comment),

Living in Granada this is an integration I look forward to use.

janiversen commented:
Lets see if CI goes green.

MartinHjelmare commented on homeassistant/components/meteoclimatic/weather.py:
```diff
@@ -0,0 +1,93 @@
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Meteoclimatic weather platform."""
+    hass_data = hass.data[DOMAIN][entry.entry_id]
```
```suggestion
    coordinator = hass.data[DOMAIN][entry.entry_id]
```

MartinHjelmare commented on homeassistant/components/meteoclimatic/weather.py:
```diff
@@ -0,0 +1,93 @@
+    """Set up the Meteoclimatic weather platform."""
+    hass_data = hass.data[DOMAIN][entry.entry_id]
+
+    async_add_entities([MeteoclimaticWeather(hass_data)], False)
```
```suggestion
    async_add_entities([MeteoclimaticWeather(coordinator)], False)
```

balloob commented on homeassistant/components/meteoclimatic/config_flow.py:
```diff
@@ -0,0 +1,65 @@
+            return self.async_abort(reason="unknown")
+
+        # Check if already configured
+        await self.async_set_unique_id(station_code)
+        self._abort_if_unique_id_configured()
```
If a user wants to set it up again, we shouldn't abort. Home Assistant will automatically remove the old config entry. 

```suggestion
        await self.async_set_unique_id(station_code, raise_on_progress=False)
```

frenck committed:
Update homeassistant/components/meteoclimatic/config_flow.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

frenck committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

frenck committed:
Remove unneeded test

frenck commented:
Applied review suggestions, removed the failing test (now obsolete).
Should be good to go now.

adrianmo commented:
Hey folks, just saw the comments and updates. Thanks a lot for taking care of it.

frenck committed:
Address pylint warning

MartinHjelmare reviewed (approved):
Looks good!

frenck commented:
🎉
